### PR TITLE
[ASDisplayNode] Renamed range update callbacks

### PR DESCRIPTION
**Pull request by garrettmoon**
We finally settled on

didEnter/ExitDisplayState
didEnter/ExitPreloadState
didEnter/ExitVisibleState

This change is meant to unify the range update methods to relate to each
other and hopefully be a bit more self explanatory.

### CODE REVIEW

File: AsyncDisplayKit/ASDisplayNode+Subclasses.h [hunk 1]
```diff
@@ -238,36 +238,58 @@ NS_ASSUME_NONNULL_BEGIN
...
-- (void)displayStateDidChange:(BOOL)inDisplayState ASDISPLAYNODE_REQUIRES_SUPER;
+- (void)didEnterDisplayState ASDISPLAYNODE_REQUIRES_SUPER;
 
 /**
- * @abstract Called whenever the the node has entered or left the load state.
+ * @abstract Called whenever the the node has exited the display state.
+ *
+ * @discussion Subclasses may use this to monitor when a node should no longer be rendering its content.
+ *
+ * @note This method can be called from any thread and should therefore be thread safe.
```
Comment (garrettmoon) [inline AsyncDisplayKit/ASDisplayNode+Subclasses.h]: All of these methods have these notes. Should they? It would be materially easier for clients if we guaranteed they were called on the main thread.

File: AsyncDisplayKit/ASDisplayNode+Subclasses.h [hunk 2]
```diff
@@ -238,36 +238,58 @@ NS_ASSUME_NONNULL_BEGIN
 - (void)interfaceStateDidChange:(ASInterfaceState)newState fromState:(ASInterfaceState)oldState ASDISPLAYNODE_REQUIRES_SUPER;
 
 /**
- * @abstract Called whenever the visiblity of the node changed.
+ * @abstract Called whenever the node becomes visible.
  *
  * @discussion Subclasses may use this to monitor when they become visible.
+ *
+ * @note This method can be called from any thread and should therefore be thread safe.
```
Comment (appleguy) [inline AsyncDisplayKit/ASDisplayNode+Subclasses.h]: @garrettmoon this one isn't right — we should make a strong guarantee that the visible state methods are called on the main thread.  It's the only way to guarantee their timing is relevant (e.g., actually called before it is visible onscreen), and work performed here is very often related to animations or other main thread things

Comment (garrettmoon) [inline AsyncDisplayKit/ASDisplayNode+Subclasses.h]: I'll update to remove, thanks!

File: AsyncDisplayKit/ASDisplayNode+Subclasses.h (hunk 1 above)
Comment (garrettmoon) [inline AsyncDisplayKit/ASDisplayNode+Subclasses.h]: My thought is we should guarantee that these will be called on main, but what does everyone else think about signing ourselves up for maintaining that?

**Comment by hannahmbanana** [comment]
When we merge this, we need to update http://asyncdisplaykit.org/docs/upgrading.html with the following:

Updated Interface State callback method names. The new names are meant to unify the range update methods to show how they relate to each other & hopefully be a bit more self explanatory:

- `-didEnter/ExitPreloadState`
- `-didEnter/ExitDisplayState`
- `-didEnter/ExitVisibleState`

These new methods replace the following:

- `-loadStateDidChange:(BOOL)inLoadState`
- `-displayStateDidChange:(BOOL)inDisplayState`
- `-visibleStateDidChange:(BOOL)isVisible`

File: AsyncDisplayKit/ASDisplayNode+Subclasses.h [hunk 3]
```diff
@@ -472,6 +494,39 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (void)asyncTraitCollectionDidChange;
 
+/**
+ * @abstract Called whenever the visiblity of the node changed.
+ *
+ * @discussion Subclasses may use this to monitor when they become visible.
+ */
+- (void)visibilityDidChange:(BOOL)isVisible ASDISPLAYNODE_REQUIRES_SUPER ASDISPLAYNODE_DEPRECATED;
```
Comment (appleguy) [inline AsyncDisplayKit/ASDisplayNode+Subclasses.h]: Are these still supposed to exist in the +Subclasses header?  Should they be in a deprecated category?

File: AsyncDisplayKit/ASDisplayNode+Subclasses.h [hunk 4]
```diff
@@ -238,36 +238,58 @@ NS_ASSUME_NONNULL_BEGIN
...
+ *
+ * @discussion Subclasses may use this to monitor data for a node should be preloaded, either from a local or remote source.
+ *
+ * @note this method is guaranteed to be called on main.
+ */
+- (void)didEnterPreloadState ASDISPLAYNODE_REQUIRES_SUPER;
+
+/**
+ * @abstract Called whenever the the node has exited the preload state.
+ *
```
Comment (appleguy) [inline AsyncDisplayKit/ASDisplayNode+Subclasses.h]: @garrettmoon I don't think it's standard in our current header docs to have extra lines between these - could you densify them a bit?  Having twice as many methods makes it harder to consume the info IMO relative to shorter block (would still be 5 full lines of comments)

File: AsyncDisplayKit/ASDisplayNode+Subclasses.h (hunk 3 above)
Comment (garrettmoon) [inline AsyncDisplayKit/ASDisplayNode+Subclasses.h]: @appleguy I can make a new ASDisplayNode+Deprecated category. Are you thinking we should move all of these there?

File: AsyncDisplayKit/ASDisplayNode+Subclasses.h (hunk 4 above)
Comment (garrettmoon) [inline AsyncDisplayKit/ASDisplayNode+Subclasses.h]: I agree that they're nicer to read without the line breaks, however it appears that there are breaks between all the '@' comments in every header I looked at (though sometimes the param ones are stuck together). Should we standardize 'em all? If so, probably best to do in a different PR?

Comment (appleguy) [inline AsyncDisplayKit/ASDisplayNode+Subclasses.h]: @garrettmoon ok, that's fine.  IIRC these were reformatted a couple months ago and it might have came in then...

File: AsyncDisplayKit/ASDisplayNode+Subclasses.h (hunk 3 above)
Comment (appleguy) [inline AsyncDisplayKit/ASDisplayNode+Subclasses.h]: @garrettmoon yeah, that seems best.  They can stay in the same file at the bottom.

File: AsyncDisplayKit.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -1881,12 +1885,15 @@
...
 			buildPhases = (
+				A5E2A5BB10F93572445AC844 /* 📦 Check Pods Manifest.lock */,
 				2E61B6A0DB0F436A9DDBE86F /* [CP] Check Pods Manifest.lock */,
 				058D09B8195D04C000B7D73C /* Sources */,
 				058D09B9195D04C000B7D73C /* Frameworks */,
 				058D09BA195D04C000B7D73C /* Resources */,
 				3B9D88CDF51B429C8409E4B6 /* [CP] Copy Pods Resources */,
 				B130AB1AC0A1E5162E211C19 /* [CP] Embed Pods Frameworks */,
+				76FDAFA9FB106E27BE016772 /* 📦 Embed Pods Frameworks */,
+				58AF99A9E1C99C30BB2F5335 /* 📦 Copy Pods Resources */,
```
Comment (garrettmoon) [inline AsyncDisplayKit.xcodeproj/project.pbxproj]: Shoot, this was added because I'm on cocoa pods 1.0. Will see if I can fix.

File: AsyncDisplayKit/ASDisplayNode+Deprecated.h [hunk 6]
```diff
@@ -0,0 +1,51 @@
+//
+//  ASDisplayNode+Deprecated.h
+//  AsyncDisplayKit
+//
+//  Created by Garrett Moon on 8/26/16.
+//  Copyright © 2016 Facebook. All rights reserved.
+//
```
Comment (maicki) [inline AsyncDisplayKit/ASDisplayNode+Deprecated.h]: Can you change the header to the default one we use in other places.

File: AsyncDisplayKit/ASDisplayNode+Deprecated.h [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+#ifndef ASDisplayNode_Deprecated_h
+#define ASDisplayNode_Deprecated_h
```
Comment (maicki) [inline AsyncDisplayKit/ASDisplayNode+Deprecated.h]: Use #pragma once

File: AsyncDisplayKit/ASDisplayNode+Subclasses.h [hunk 8]
```diff
@@ -238,36 +238,58 @@ NS_ASSUME_NONNULL_BEGIN
 - (void)interfaceStateDidChange:(ASInterfaceState)newState fromState:(ASInterfaceState)oldState ASDISPLAYNODE_REQUIRES_SUPER;
 
 /**
- * @abstract Called whenever the visiblity of the node changed.
+ * @abstract Called whenever the node becomes visible.
  *
  * @discussion Subclasses may use this to monitor when they become visible.
+ *
+ * @note this method is guaranteed to be called on main.
```
Comment (maicki) [inline AsyncDisplayKit/ASDisplayNode+Subclasses.h]: Should we start with uppercase the "this ..."

File: AsyncDisplayKit/ASDisplayNode.mm [hunk 9]
```diff
@@ -2432,26 +2432,38 @@ - (void)recursivelyClearFetchedData
   });
 }
 
-- (void)visibilityDidChange:(BOOL)isVisible
+- (void)didEnterVisibleState
 {
-  // subclass override
+  //subclass override
```
Comment (maicki) [inline AsyncDisplayKit/ASDisplayNode.mm]: Why remove the " "?

Comment (garrettmoon) [inline AsyncDisplayKit/ASDisplayNode.mm]: Was just trying to match them up. I can add one in all of them instead?

File: AsyncDisplayKit/ASNetworkImageNode.mm [hunk 10]
```diff
@@ -281,21 +281,31 @@ - (void)displayWillStart
...
  in ASMultiplexImageNode as well. */
-- (void)visibleStateDidChange:(BOOL)isVisible
+- (void)didEnterVisibleState
 {
-  [super visibleStateDidChange:isVisible];
-
+  [super didEnterVisibleState];
+  
   if (_downloaderFlags.downloaderImplementsSetPriority) {
     ASDN::MutexLocker l(__instanceLock__);
```
Comment (maicki) [inline AsyncDisplayKit/ASNetworkImageNode.mm]: If we access the `_downloaderFlags` do we not also need to hold the `__instanceLock__ `?. Also in `didExitVisibleState`

File: AsyncDisplayKit/ASMultiplexImageNode.mm [hunk 11]
```diff
@@ -293,20 +293,30 @@ - (void)displayWillStart
...
+/* didEnterVisibleState / didExitVisibleState in ASNetworkImageNode has a very similar implementation. Changes here are likely necessary
  in ASNetworkImageNode as well. */
-- (void)visibleStateDidChange:(BOOL)isVisible
+- (void)didEnterVisibleState
 {
-  [super visibleStateDidChange:isVisible];
+  [super didEnterVisibleState];
   
   if (_downloaderImplementsSetPriority) {
     ASDN::MutexLocker l(_downloadIdentifierLock);
```
Comment (maicki) [inline AsyncDisplayKit/ASMultiplexImageNode.mm]: Do we not hold the lock while accessing `_downloaderImplementsSetPriority`?

File: AsyncDisplayKit/ASNetworkImageNode.mm (hunk 10 above)
Comment (garrettmoon) [inline AsyncDisplayKit/ASNetworkImageNode.mm]: _downloaderFlags is only ever set on init.

File: AsyncDisplayKit/ASMultiplexImageNode.mm (hunk 11 above)
Comment (garrettmoon) [inline AsyncDisplayKit/ASMultiplexImageNode.mm]: Also only set on init.

**Comment by Adlai-Holler** [comment]
💯

**Comment by maicki** [comment]
LGTM

**Comment by Adlai-Holler** [comment]
Travis failure is spurious. Landing
